### PR TITLE
Added MergeTickets functionality and tests - KLK

### PR DESCRIPTION
**Pull request by klkitchens**
Hope I'm doing this correctly.  Have not contributed before outside of work.

As I asked on the gitterhub forum, added the MergeTickets API functionality as well as tests for validation.

If you need more information, please let me know what to do :)

### CODE REVIEW

**Comment by klkitchens** [comment]
Those tests are failing in your branch too, no?  I didn't touch any of that code in the failing areas.

File: src/ZendeskApi_v2/Requests/Tickets.cs [hunk 1]
```diff
@@ -470,6 +473,22 @@ public bool DeleteMultiple(IEnumerable<long> ids)
...
+        /// <param name="sourceTicketIds">List of ids of source tickets to be merged from.</param>
+        /// <param name="targetComment">Private comment to add to the target ticket (optional but recommended)</param>
+        /// <param name="sourceComment">Private comment to add to the source ticket(s) (optional but recommended)</param>
+        /// <param name="targetCommentPublic">Whether comment in target ticket is public or private (default = private)</param>
+        /// <param name="sourceCommentPublic">Whether comment in source ticket is public or private (default = private)</param>
+        /// <returns></returns>
+        public JobStatusResponse MergeTickets(long targetTicketId, IEnumerable<long> sourceTicketIds, string targetComment = "", string sourceComment = "", bool targetCommentPublic = false, bool sourceCommentPublic = false)
+        {
+            return GenericPost<JobStatusResponse>($"{_tickets}/{targetTicketId}/merge.json", new { ids = sourceTicketIds.ToArray(), target_comment = targetComment, source_comment = sourceComment, target_comment_is_public = targetCommentPublic, source_comment_is_public = sourceCommentPublic });
```
Comment (mozts2005) [inline src/ZendeskApi_v2/Requests/Tickets.cs:488]: ToArray is not needed.

```suggestion
            return GenericPost<JobStatusResponse>($"{_tickets}/{targetTicketId}/merge.json", new { ids = sourceTicketIds, target_comment = targetComment, source_comment = sourceComment, target_comment_is_public = targetCommentPublic, source_comment_is_public = sourceCommentPublic });
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 2]
```diff
@@ -1119,6 +1119,84 @@ public void CanImportTicketAsync()
...
+            var targetMergeComment = $"Merged with tickets {string.Join(", ", mergeIds.ToArray())}";
+            var sourceMergeComment = $"Closing in favor of {targetTicketId}";
+
+            var res = api.Tickets.MergeTickets(
+                targetTicketId, 
+                mergeIds, 
+                targetMergeComment, 
+                sourceMergeComment);
+
+            Assert.AreEqual(res.JobStatus.Status, "queued");
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1162]:
```suggestion
            Assert.That(res.JobStatus.Status, Is.EqualTo("queued"));
```
please change all Asserts to use the ```Assert.That``` style.

File: src/ZendeskApi_v2/Requests/Tickets.cs [hunk 3]
```diff
@@ -470,6 +473,22 @@ public bool DeleteMultiple(IEnumerable<long> ids)
...
+        /// Merges the source tickets in the "ids" list into the target ticket with comments as defined.
+        /// </summary>
+        /// <param name="targetTicketId">Id of ticket to be merged into.</param>
+        /// <param name="sourceTicketIds">List of ids of source tickets to be merged from.</param>
+        /// <param name="targetComment">Private comment to add to the target ticket (optional but recommended)</param>
+        /// <param name="sourceComment">Private comment to add to the source ticket(s) (optional but recommended)</param>
+        /// <param name="targetCommentPublic">Whether comment in target ticket is public or private (default = private)</param>
+        /// <param name="sourceCommentPublic">Whether comment in source ticket is public or private (default = private)</param>
+        /// <returns></returns>
+        public JobStatusResponse MergeTickets(long targetTicketId, IEnumerable<long> sourceTicketIds, string targetComment = "", string sourceComment = "", bool targetCommentPublic = false, bool sourceCommentPublic = false)
```
Comment (mozts2005) [inline src/ZendeskApi_v2/Requests/Tickets.cs:486]: Please add both the Sync and Async version of the new method.

File: src/ZendeskApi_v2/Requests/Tickets.cs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Linq;
```
Comment (mozts2005) [inline src/ZendeskApi_v2/Requests/Tickets.cs:3]: Not needed

File: test/ZendeskApi_v2.Test/TicketTests.cs (hunk 2 above)
Comment (klkitchens) [reply test/ZendeskApi_v2.Test/TicketTests.cs:1162]: Changing... used .AreEqual based on your existing code. :)

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 5]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+            var comments = api.Tickets.GetTicketComments(targetTicketId);
+            Assert.That(comments.Comments.Count, Is.EqualTo(2));
+            Assert.That(comments.Comments[0].Body, Is.EqualTo(targetDescription));
+            Assert.That(comments.Comments[1].Body, Is.EqualTo(targetMergeComment));
+
+            api.Tickets.DeleteAsync(targetTicketId);
+        }
+
+        [Test]
+        public void CanMergeTicketsAsync()
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1201]:
```suggestion
        public async Task CanMergeTicketsAsync()
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 6]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+            };
+            var targetTicket = new Ticket
+            {
+                Subject = "Target Ticket",
+                Comment = new Comment { Body = targetDescription, Public = true, }
+            };
+
+            var mergeIds = new List<long>();
+
+            var tick = api.Tickets.CreateTicket(sourceTicket1);
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1224]:
```suggestion
            var tick = await api.Tickets.CreateTicketAsync(sourceTicket1);
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 7]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+            mergeIds.Add(tick.Ticket.Id.Value);
+            tick = api.Tickets.CreateTicket(sourceTicket2);
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1226]:
```suggestion
            tick = await api.Tickets.CreateTicketAsync(sourceTicket2);
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 8]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+            mergeIds.Add(tick.Ticket.Id.Value);
+            tick = api.Tickets.CreateTicket(targetTicket);
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1228]:
```suggestion
            tick = await api.Tickets.CreateTicketAsync(targetTicket);
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 9]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+            var targetTicketId = tick.Ticket.Id.Value;
+
+            var targetMergeComment = $"Merged with tickets {string.Join(", ", mergeIds.ToArray())}";
+            var sourceMergeComment = $"Closing in favor of {targetTicketId}";
+
+            var res = api.Tickets.MergeTicketsAsync(
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1234]:
```suggestion
            var res = await api.Tickets.MergeTicketsAsync(
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 10]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+                targetTicketId,
+                mergeIds,
+                targetMergeComment,
+                sourceMergeComment);
+
+            Assert.That(res.Result.JobStatus.Status, Is.EqualTo("queued"));
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1240]:
```suggestion
            Assert.That(res.JobStatus.Status, Is.EqualTo("queued"));
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 11]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+
+            do
+            {
+                Thread.Sleep(5000);
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1244]:
```suggestion
               await Task.Delay(5000);
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 12]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+                var job = api.JobStatuses.GetJobStatus(res.Result.JobStatus.Id);
+                Assert.That(job.JobStatus.Id, Is.EqualTo(res.Result.JobStatus.Id));
+
+                if (job.JobStatus.Status == "completed") break;
+            } while (true);
+
+            var counter = 0;
+            foreach (var id in mergeIds)
+            {
+                var oldTicket = api.Tickets.GetTicket(id);
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1254]:
```suggestion
                var oldTicket = await api.Tickets.GetTicketAsync(id);
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 13]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+                Assert.That(oldTicket.Ticket.Id.Value, Is.EqualTo(id));
+                Assert.That(oldTicket.Ticket.Status, Is.EqualTo("closed"));
+
+                var oldComments = api.Tickets.GetTicketComments(id);
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1258]:
```suggestion
                var oldComments = await api.Tickets.GetTicketCommentsAsync(id);
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 14]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+                Assert.That(oldComments.Comments.Count, Is.EqualTo(2));
+                Assert.That(oldComments.Comments[0].Body, Is.EqualTo(sourceDescription[counter]));
+                Assert.That(oldComments.Comments[1].Body, Is.EqualTo(sourceMergeComment));
+
+                api.Tickets.DeleteAsync(id);
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1263]:
```suggestion
              await api.Tickets.DeleteAsync(id);
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 15]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+                counter++;
+            }
+
+            var ticket = api.Tickets.GetTicket(targetTicketId);
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1267]:
```suggestion
            var ticket = await api.Tickets.GetTicketAsync(targetTicketId);
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 16]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+            Assert.That(ticket.Ticket.Id.Value, Is.EqualTo(targetTicketId));
+
+            var comments = api.Tickets.GetTicketComments(targetTicketId);
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1270]:
```suggestion
            var comments =await api.Tickets.GetTicketCommentsAsync(targetTicketId);
```

File: test/ZendeskApi_v2.Test/TicketTests.cs [hunk 17]
```diff
@@ -1119,6 +1119,162 @@ public void CanImportTicketAsync()
...
+            Assert.That(comments.Comments.Count, Is.EqualTo(2));
+            Assert.That(comments.Comments[0].Body, Is.EqualTo(targetDescription));
+            Assert.That(comments.Comments[1].Body, Is.EqualTo(targetMergeComment));
+
+            api.Tickets.DeleteAsync(targetTicketId);
```
Comment (mozts2005) [inline test/ZendeskApi_v2.Test/TicketTests.cs:1275]:
```suggestion
           await api.Tickets.DeleteAsync(targetTicketId);
```

File: test/ZendeskApi_v2.Test/TicketTests.cs (hunk 6 above)
Comment (klkitchens) [reply test/ZendeskApi_v2.Test/TicketTests.cs:1224]: Did not seem any reason to call everything else Async just to test the Merge Async call.  Again, followed the existing code as a guideline.

Comment (mozts2005) [reply test/ZendeskApi_v2.Test/TicketTests.cs:1224]: Can you please share where you see that an async is only using async for part of the test?

**Comment by klkitchens** [comment]
public void CanImportTicketAsync()

var resComments = api.Tickets.GetTicketComments(res.Result.Ticket.Id.Value);

Also using the old Assert logic too.

I didn't search, just happened to be the first one above my code. :)

**Comment by mozts2005** [comment]
Thanks for sharing that I was under the impression that I had fixed all of them. 
I will make a point of updating that one.

**Comment by mozts2005** [comment]
So as soon as I have the Async fixes I think we can merge this addition.

**Comment by klkitchens** [comment]
> 
> 
> So as soon as I have the Async fixes I think we can merge this addition.

And how long to a release?  I can use my own dll, but would prefer to keep the official nuget version.

**Comment by mozts2005** [comment]
Should be able to have it out later today.

**Comment by klkitchens** [comment]
> 
> 
> Should be able to have it out later today. As long as you have the updated Async test added to your PR.

It's in my repo, but my repo is not sending that last change over...

**Comment by mozts2005** [comment]
Try closing this PR and open a new one.

**Pull request closed without merging**